### PR TITLE
Better overlap and gap handling

### PR DESCRIPTION
**Pull request by kayoub5**
Handle the case were parameters overlap and have gap between them a bit better

Currently the encoding logic do encode them correctly but still bails if it encounter such parameters due to bit_length checking

One simple example is a struct defining a DTC code, it can be treated as a whole number, or have its individual components each treated as a parameter, for convenience, user could define both in the same struct.

image

### CODE REVIEW

**Comment by kayoub5** [comment]
@andlaus review?

File: odxtools/structures.py [hunk 1]
```diff
@@ -122,14 +125,60 @@ def convert_physical_to_internal(self,
...
+
+    def _get_encode_error_str(self,
+            verb: str,
+            coded_rpc: bytearray,
+            bit_length: int):
+
+        return str(f"Structure {self.short_name} {verb} encoded uncorrectly:" +
+                f" actual length is {len(coded_rpc)}," +
+                f" computed byte length is {bit_length // 8}," +
+                f" computed_rpc is {coded_rpc.hex()}" +
```
Comment (kayoub5) [inline odxtools/structures.py:178]:
```suggestion
                f" computed_rpc is {coded_rpc.hex()}\n" +
```

File: odxtools/structures.py [hunk 2]
```diff
@@ -122,14 +125,60 @@ def convert_physical_to_internal(self,
...
 
         return bytearray(coded_rpc)
 
+    def _validate_coded_rpc(
+            self,
+            coded_rpc: bytearray,
+            bit_length: Optional[int]):
+
+        if self._byte_size is not None:
+            # We definetly broke something if we didn't respect the explicit
```
Comment (andlaus) [inline odxtools/structures.py:143]:
```suggestion
            # We definitely broke something if we didn't respect the explicit
```

**Review by andlaus — COMMENTED** [review]
looks good. If possible, I would like to see the somersault example to be extended by such a case, though...

**Comment by andlaus** [comment]
> I have limited time at the moment for writing detailed tests, I could try to add those afterwards (different PR)

ok, let's get this merged. be aware though, that not having this in the example PDX, fixing issues like #70 is difficult for people who don't have datasets that use this feature.

> Tip: you may want to add Codecov https://about.codecov.io/ to the CI, it would help with telling what changes requires tests and what changes are already covered with existing tests.

Yep, this has been on my TODO list for quite a while, but more pressing matters kept interfering and even my time is limited ;)

maybe soon.

**Comment by kayoub5** [comment]
@andlaus could you publish a version with the PRs merged so far?

**Comment by andlaus** [comment]
np. done.

**Pull request merged**